src/App.js

```diff
@@ -16,60 +16,92 @@ import apiEndpoints from "./config";
-
-  const handleItemSelected = async (itemName, columnIndex) => {
-    // Start building the path from 'published'
-    let path = "published";
-
-    // Update columns up to the current index
-    const updatedColumns = columns.slice(0, columnIndex + 1);
-    updatedColumns[columnIndex] = {
-      ...updatedColumns[columnIndex],
-      items: updatedColumns[columnIndex].items.map((item) => ({
-        ...item,
-        isSelected: item.name === itemName,
-      })),
-      selected: itemName,
-    };
-
-    // Include the selected items from previous columns in the path
-    updatedColumns.forEach((col) => {
-      if (col.selected) {
-        path += `.${col.selected}`;
-      }
+    // Fetch columns
+    const columnPromises = columnPaths.map((path, idx) =>
+      columns[idx].items.length === 0 ? fetchChildren(apiEndpoint, path) : null
+    );
+    Promise.all(columnPromises).then((responses) => {
+      setColumns((prevColumns) =>
+        prevColumns.map((column, idx) => ({
+          selected: column.selected,
+          items: responses[idx]
+            ? responses[idx].map((name) => ({
+                name,
+                isSelected: prevColumns[idx + 1]?.selected === name,
+              }))
+            : column.items,
+        }))
+      );
     });
 
-    const children = await fetchChildren(apiEndpoint, path);
-    if (children.length === 0) {
-      const data = await fetchData(apiEndpoint, path);
+    // Fetch data
+    fetchData(apiEndpoint, columnPaths.at(-1)).then((data) => {
       setDataView(cleanJSON(JSON.stringify(data), null, 2));
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [apiEndpoint, path]);
+
+  useEffect(() => {
+    const pathString = columns
+      .slice(1)
+      .map((col) => col.selected)
+      .join(".");
+    const newSearchParams = new URLSearchParams(window.location.search);
+    newSearchParams.set("path", pathString);
+
+    if (window.history.pushState) {
+      const newurl =
+        window.location.protocol +
+        "//" +
+        window.location.host +
+        window.location.pathname +
+        "?" +
+        newSearchParams.toString();
+      window.history.pushState({ path: newurl }, "", newurl);
+      setPath(pathString);
     }
-    const newColumns = updatedColumns.slice(0, columnIndex + 1);
-    newColumns.push({
-      items: children.map((name) => ({ name, isSelected: false })),
-      selected: null,
+  }, [columns]);
+
+  const handleItemSelected = async (itemName, columnIndex) => {
+    // return if the item is already selected
+    if (columns[columnIndex + 1]?.selected === itemName) return;
+
+    setDataView("");
+    setColumns((prevColumns) => {
+      const newColumns = prevColumns.slice(0, columnIndex + 1);
+      newColumns[columnIndex] = {
+        ...newColumns[columnIndex],
+        items: newColumns[columnIndex].items.map((item) => ({
+          ...item,
+          isSelected: item.name === itemName,
+        })),
+      };
```

Comment:
I think `setColumns` contains too much inline logic. More appropriate to create a separate function for clarity.

Reply:
i think its clear as is. creating a seperate function would just move the same logic to some place